script.module.trakt/lib/trakt/interfaces/oauth/device.py

```diff
@@ -0,0 +1,174 @@
+
+        response = self.http.post(
+            'token',
+            data={
+                'client_id': client_id,
+                'client_secret': client_secret,
+
+                'code': device_code
+            }
+        )
+
+        data = self.get_data(response, **kwargs)
+
+        if not data:
+            return None
+
+        return data
+
+
+class DeviceOAuthPoller(Interface, Emitter):
+    def __init__(self, client, device_code, expires_in, interval):
+        super(DeviceOAuthPoller, self).__init__(client)
+
+        self.device_code = device_code
+        self.expires_in = expires_in
+        self.interval = interval
+
+        # Calculate code expiry date/time
+        self.expires_at = datetime.utcnow() + timedelta(seconds=self.expires_in)
+
+        # Private attributes
+        self._abort = False
+        self._active = False
+        self._running = False
+        self._thread = None
+
+    @property
+    def active(self):
+        return self._active
+
+    def has_expired(self):
+        return datetime.utcnow() > self.expires_at
+
+    def start(self, daemon=None):
+        if self._active or self._thread:
+            raise Exception('Poller already started')
+
+        # Construct thread process wrapper
+        def wrapper():
+            try:
+                self._process()
+            except Exception as ex:
+                log.warn('Exception raised in DeviceOAuthPoller: %s', ex, exc_info=True)
+            finally:
+                self._active = False
+                self._running = False
+
+                if self._abort:
+                    self.emit('aborted')
+
+        # Construct poller thread
+        self._thread = Thread(
+            target=wrapper,
+            name='%s:%s' % (DeviceOAuthPoller.__module__, DeviceOAuthPoller.__name__)
+        )
+
+        # Set `daemon` state
+        if daemon is not None:
+            self._thread.daemon = daemon
+
+        # Start polling
+        self._abort = False
+        self._active = True
+        self._running = True
+        self._thread.start()
+
+    def stop(self):
+        # Flag as thread abort
+        self._abort = True
+
+        # Flag thread to stop
+        self._running = False
+
+    def _process(self):
+        while self._running:
+            # Ensure code hasn't expired yet
+            if self.has_expired():
+                self.emit('expired')
+                break
+
+            # Trigger "poll" event, check if we should continue polling
+            if not self._should_poll():
+                self.stop()
+                break
+
+            # Poll for token
+            response = self.client['oauth/device'].token(self.device_code, parse=False)
+
+            if response:
+                # Authentication complete
+                self.emit('authenticated', self.get_data(response))
+                break
+
+            # Sleep for defined interval
+            time.sleep(self.interval)
```

Comment:
Is there a reason for using time.sleep, rather than xbmc.sleep here? Timing looks like it's in secs so xbmc.Monitor().waitForAbort might be an even better choice.